plutus-core/plutus-core/src/PlutusCore/Parser/Type.hs

```diff
@@ -58,15 +70,57 @@ pType = choice $ map try
     , varType
     ]
 
-defaultUniType :: Parser (SomeTypeIn DefaultUni)
-defaultUniType = choice $ map try
-  [ inParens defaultUniType
-  , SomeTypeIn DefaultUniInteger <$ symbol "integer"
-  , SomeTypeIn DefaultUniByteString <$ symbol "bytestring"
-  , SomeTypeIn DefaultUniString <$ symbol "string"
-  , SomeTypeIn DefaultUniUnit <$ symbol "unit"
-  , SomeTypeIn DefaultUniBool <$ symbol "bool"
-  ] --TODO complete all defaultUni types
+-- | Parser for built-in type applications.
+defaultUniApplication :: Parser (SomeTypeIn (Kinded DefaultUni))
+defaultUniApplication = do
+    -- Parse the head of the application.
+    f <- defaultUni
+    -- Parse the arguments.
+    as <- many defaultUni
+    -- Iteratively apply the head to the arguments checking that the kinds match and
+    -- failing otherwise.
+    foldM tryUniApply f as
+
+-- | Parser for built-in types (the ones from 'DefaultUni' specifically).
+--
+-- 'Kinded' is needed for checking that a type function can be applied to its argument.
+-- I.e. we do Plutus kind checking of builtin type applications during parsing, which is
```

Comment:
Who says we don't have an intrinsically typed AST :laughing:

Reply:
Only fools!